File: svws-webclient/src/ts-lib/ts/src/core/abschluss/gost/belegpruefung/Schwerpunkt.ts
import { JavaObject, cast_java_lang_Object } from '../../../../java/lang/JavaObject';
import { Naturwissenschaften, cast_de_nrw_schule_svws_core_abschluss_gost_belegpruefung_Naturwissenschaften } from '../../../../core/abschluss/gost/belegpruefung/Naturwissenschaften';
import { Fremdsprachen, cast_de_nrw_schule_svws_core_abschluss_gost_belegpruefung_Fremdsprachen } from '../../../../core/abschluss/gost/belegpruefung/Fremdsprachen';
import { GostBelegpruefungsArt, cast_de_nrw_schule_svws_core_abschluss_gost_GostBelegpruefungsArt } from '../../../../core/abschluss/gost/GostBelegpruefungsArt';
import { GostBelegpruefung, cast_de_nrw_schule_svws_core_abschluss_gost_GostBelegpruefung } from '../../../../core/abschluss/gost/GostBelegpruefung';
import { AbiturdatenManager, cast_de_nrw_schule_svws_core_abschluss_gost_AbiturdatenManager } from '../../../../core/abschluss/gost/AbiturdatenManager';
import { GostBelegungsfehler, cast_de_nrw_schule_svws_core_abschluss_gost_GostBelegungsfehler } from '../../../../core/abschluss/gost/GostBelegungsfehler';

export class Schwerpunkt extends GostBelegpruefung {


	/**
	 * Erstellt eine neue Belegprüfung für den Schwerpunkt.
	 * 
	 * @param manager             der Daten-Manager für die Abiturdaten
	 * @param pruefungs_art       die Art der durchzuführenden Prüfung (z.B. EF.1 oder GESAMT)
	 * @param pruefung_sprachen   das Ergebnis für die Belegprüfung der Sprachen
	 * @param pruefung_nawi       das Ergebnis für die Belegprüfung der Naturwissenschaften
	 */
	public constructor(manager : AbiturdatenManager, pruefungs_art : GostBelegpruefungsArt, pruefung_sprachen : Fremdsprachen, pruefung_nawi : Naturwissenschaften) {
		super(manager, pruefungs_art, pruefung_sprachen, pruefung_nawi);
	}

	protected init() : void {
	}

	protected pruefeEF1() : void {
		let pruefung_sprachen : Fremdsprachen = (cast_de_nrw_schule_svws_core_abschluss_gost_belegpruefung_Fremdsprachen(this.pruefungen_vorher[0]));
		let pruefung_nawi : Naturwissenschaften = (cast_de_nrw_schule_svws_core_abschluss_gost_belegpruefung_Naturwissenschaften(this.pruefungen_vorher[1]));
		if ((pruefung_sprachen.getAnzahlDurchgehendSchritflichBelegt() >= 2) && (pruefung_nawi.getAnzahlDurchgehendBelegt() >= 2) && (pruefung_nawi.getAnzahlDurchgehendSchritflichBelegt() >= 1)) 
			return;
		if (pruefung_sprachen.getAnzahlDurchgehendSchritflichBelegt() >= 2) {
			this.addFehler(GostBelegungsfehler.NW_FS_12_INFO);
			return;
		}
		if ((pruefung_nawi.getAnzahlDurchgehendBelegt() >= 2) && (pruefung_nawi.getAnzahlDurchgehendSchritflichBelegt() >= 1)) {
			this.addFehler(GostBelegungsfehler.NW_FS_13_INFO);
			return;
		}
		this.addFehler(GostBelegungsfehler.NW_FS_10);
	}

	protected pruefeGesamt() : void {
		let pruefung_sprachen : Fremdsprachen = (cast_de_nrw_schule_svws_core_abschluss_gost_belegpruefung_Fremdsprachen(this.pruefungen_vorher[0]));
		let pruefung_nawi : Naturwissenschaften = (cast_de_nrw_schule_svws_core_abschluss_gost_belegpruefung_Naturwissenschaften(this.pruefungen_vorher[1]));
		if ((pruefung_sprachen.getAnzahlDurchgehendSchritflichBelegt() >= 2) && (pruefung_nawi.getAnzahlDurchgehendBelegt() >= 2) && (pruefung_nawi.getAnzahlDurchgehendSchritflichBelegt() >= 1)) 
			return;
		if (pruefung_sprachen.getAnzahlDurchgehendSchritflichBelegt() >= 2) {
			this.addFehler(GostBelegungsfehler.NW_FS_12_INFO);
			return;
		}
		if ((pruefung_nawi.getAnzahlDurchgehendBelegt() >= 2) && (pruefung_nawi.getAnzahlDurchgehendSchritflichBelegt() >= 1)) {
			this.addFehler(GostBelegungsfehler.NW_FS_13_INFO);
			return;
		}
		this.addFehler(GostBelegungsfehler.NW_FS_10);
	}

	isTranspiledInstanceOf(name : string): boolean {
		return ['de.nrw.schule.svws.core.abschluss.gost.GostBelegpruefung', 'de.nrw.schule.svws.core.abschluss.gost.belegpruefung.Schwerpunkt'].includes(name);
	}

}

export function cast_de_nrw_schule_svws_core_abschluss_gost_belegpruefung_Schwerpunkt(obj : unknown) : Schwerpunkt {
	return obj as Schwerpunkt;
}

File: svws-core/src/main/java/de/nrw/schule/svws/core/abschluss/gost/belegpruefung/Schwerpunkt.java
package de.nrw.schule.svws.core.abschluss.gost.belegpruefung;

import de.nrw.schule.svws.core.abschluss.gost.AbiturdatenManager;
import de.nrw.schule.svws.core.abschluss.gost.GostBelegpruefung;
import de.nrw.schule.svws.core.abschluss.gost.GostBelegpruefungsArt;
import de.nrw.schule.svws.core.abschluss.gost.GostBelegungsfehler;
import jakarta.validation.constraints.NotNull;

/**
 * Diese Klasse enthält die Belegprüfungen bezüglich eines möglichen Schwerpunktes eines Schüler 
 * im naturwissenschaftlichen oder sprachlichen Bereich für die Prüfung der EF1 bzw. 
 * für die Gesamtprüfungen.
 */
public class Schwerpunkt extends GostBelegpruefung {

	/**
	 * Erstellt eine neue Belegprüfung für den Schwerpunkt.
	 * 
	 * @param manager             der Daten-Manager für die Abiturdaten
	 * @param pruefungs_art       die Art der durchzuführenden Prüfung (z.B. EF.1 oder GESAMT)
	 * @param pruefung_sprachen   das Ergebnis für die Belegprüfung der Sprachen
	 * @param pruefung_nawi       das Ergebnis für die Belegprüfung der Naturwissenschaften
	 */	
	public Schwerpunkt(final @NotNull AbiturdatenManager manager, final @NotNull GostBelegpruefungsArt pruefungs_art,
	final @NotNull Fremdsprachen pruefung_sprachen, final @NotNull Naturwissenschaften pruefung_nawi) {
		super(manager, pruefungs_art, pruefung_sprachen, pruefung_nawi);
	}


	@Override
	protected void init() {
		// Keine Initialisierung notwendig
	}


	
	
	@Override
	protected void pruefeEF1() {
		final @NotNull Fremdsprachen pruefung_sprachen = ((@NotNull Fremdsprachen)pruefungen_vorher[0]);
		final @NotNull Naturwissenschaften pruefung_nawi = ((@NotNull Naturwissenschaften)pruefungen_vorher[1]);
		
		// Prüfe, ob insgesamt so viele Fremdsprachen und Naturwissenschaften gewählt wurden, dass zunächst kein Schwerpunkt vorliegt.
		// Dann liegt kein Belegungsfehler vor.
		if ((pruefung_sprachen.getAnzahlDurchgehendSchritflichBelegt() >= 2) &&
		    (pruefung_nawi.getAnzahlDurchgehendBelegt() >= 2) && (pruefung_nawi.getAnzahlDurchgehendSchritflichBelegt() >= 1))
			return;
 
		// Prüfe, ob ein sprachlicher Schwerpunkt vorliegt
		if (pruefung_sprachen.getAnzahlDurchgehendSchritflichBelegt() >= 2) {
			addFehler(GostBelegungsfehler.NW_FS_12_INFO);
			return;
		}

		// Prüfe, ob ein naturwissenschaftlicher Schwerpunkt vorliegt
		if ((pruefung_nawi.getAnzahlDurchgehendBelegt() >= 2) && (pruefung_nawi.getAnzahlDurchgehendSchritflichBelegt() >= 1)) {
			addFehler(GostBelegungsfehler.NW_FS_13_INFO);
			return;
		}

		// Es wurden zu wenig Fremdsprachen und Naturwissenschaften belegt -> Belegungsfehler
		addFehler(GostBelegungsfehler.NW_FS_10);
	}


	
	
	@Override
	protected void pruefeGesamt() {
		final @NotNull Fremdsprachen pruefung_sprachen = ((@NotNull Fremdsprachen)pruefungen_vorher[0]);
		final @NotNull Naturwissenschaften pruefung_nawi = ((@NotNull Naturwissenschaften)pruefungen_vorher[1]);
		
		// Prüfe, ob insgesamt so viele Fremdsprachen und Naturwissenschaften gewählt wurden, dass zunächst kein Schwerpunkt vorliegt.
		// Dann liegt kein Belegungsfehler vor.
		if ((pruefung_sprachen.getAnzahlDurchgehendSchritflichBelegt() >= 2) &&
		    (pruefung_nawi.getAnzahlDurchgehendBelegt() >= 2) && (pruefung_nawi.getAnzahlDurchgehendSchritflichBelegt() >= 1))
			return;
 
		// Prüfe, ob ein sprachlicher Schwerpunkt vorliegt
		if (pruefung_sprachen.getAnzahlDurchgehendSchritflichBelegt() >= 2) {
			addFehler(GostBelegungsfehler.NW_FS_12_INFO);
			return;
		}

		// Prüfe, ob ein naturwissenschaftlicher Schwerpunkt vorliegt
		if ((pruefung_nawi.getAnzahlDurchgehendBelegt() >= 2) && (pruefung_nawi.getAnzahlDurchgehendSchritflichBelegt() >= 1)) {
			addFehler(GostBelegungsfehler.NW_FS_13_INFO);
			return;
		}

		// Es wurden zu wenig Fremdsprachen und Naturwissenschaften belegt -> Belegungsfehler
		addFehler(GostBelegungsfehler.NW_FS_10);
	}

}
